### DISCUSSION TITLE
How to explain why `startinsert()` does not work?

### DISCUSSION THREAD

**Discussion by przepompownia** [willothy/flatten.nvim#81] [Q&A] [answered]
Hi, please look at the following `init.lua` (to run in some new directory with `.../path/to/nvim --clean -u init.lua`)

```lua
local thisInitFile = debug.getinfo(1).source:match('@?(.*)')
local configDir = vim.fs.dirname(thisInitFile)
local exepath = vim.uv.exepath()

vim.env['XDG_CONFIG_HOME'] = configDir
vim.env['XDG_DATA_HOME'] = configDir .. '/.xdg/data'
vim.env['XDG_STATE_HOME'] = configDir .. '/.xdg/state'
vim.env['XDG_CACHE_HOME'] = configDir .. '/.xdg/cache'
local stdPathConfig = vim.fn.stdpath('config')

vim.opt.runtimepath:prepend(stdPathConfig)
vim.opt.packpath:prepend(stdPathConfig)
local pluginsPath = vim.fn.fnamemodify('plugins/', ':p')

for name, url in pairs{
  ['flatten-nvim'] = 'https://github.com/willothy/flatten.nvim',
} do
  local installPath = pluginsPath .. name
  if vim.fn.isdirectory(installPath) == 0 then
    vim.fn.system {
      'git',
      'clone',
      '--depth=1',
      url,
      installPath,
    }
  end
  vim.opt.runtimepath:append(installPath)
end

require('flatten').setup({
  callbacks = {
    post_open = function (bufnr, winnr, ft, _is_blocking)
      if ft == 'gitcommit' then
        vim.wo[winnr].number = true
        vim.cmd.startinsert()
        vim.api.nvim_create_autocmd({'QuitPre'}, {
          buffer = bufnr,
          once = true,
          callback = vim.schedule_wrap(function ()
            vim.api.nvim_buf_delete(bufnr, {})
          end),
        })
      end
    end,
  }
})
vim.api.nvim_create_autocmd('FileType', {
  pattern = 'gitcommit',
  callback = function ()
    vim.cmd.startinsert()
  end,
})
vim.api.nvim_create_autocmd('UIEnter', {
  callback = function ()
    vim.cmd.new({mods = {split = 'botright'}})
    vim.fn.termopen(
      vim.go.shell,
      {
        env = {
          GIT_EDITOR = table.concat({
            exepath,
            '--clean',
            '-u',
            thisInitFile
          }, ' ')
        },
      }
    )
  end,
})
```

It does not work both in `post_open` and `FileType` autocommand. This doesn't cause me any problem to use, but I can't explain this behavior. Any idea?

![flatten](https://github.com/willothy/flatten.nvim/assets/11404453/eff306cc-575a-43aa-8e80-db353e893384)

**Comment by willothy** [comment]
I am not sure, but it is possible that because the hooks are executed in a function that is called via RPC, the API is unavailable / won't work. Have you tried wrapping the hook in `vim.schedule_wrap`?

Note: Other hooks should not be scheduled as they are expected to be executed synchronously, but the post_open can be non-blocking if needed.

**Comment by przepompownia** [comment]
```lua
require('flatten').setup({
  callbacks = {
    post_open = function (bufnr, winnr, ft, _is_blocking)
      if ft == 'gitcommit' then
        vim.wo[winnr].number = true
        vim.schedule_wrap(function()
          vim.cmd.startinsert()
        end)
        vim.api.nvim_create_autocmd({'QuitPre'}, {
          buffer = bufnr,
          once = true,
          callback = vim.schedule_wrap(function ()
            vim.api.nvim_buf_delete(bufnr, {})
          end),
        })
      end
    end,
  }
})
vim.api.nvim_create_autocmd('FileType', {
  pattern = 'gitcommit',
  callback = vim.schedule_wrap(function ()
    vim.cmd.startinsert()
  end),
})
```
Do you mean something like above? I tried also wrapping the whole `post_open` but nothing new happened. 

Anyway, thanks for the plugin as an interesting example of what we can do with Neovim.

**Reply by willothy** [reply]
If you're calling it like that then use `vim.schedule`, `vim.schedule_wrap` is equivalent to 

```lua 
function(f) 
  return function() 
    vim.schedule(f)
  end
end
```

But if wrapping the function didn't work, then nevermind, I'm not sure. I'll look into it more though. Glad you're enjoying the plugin!

**Comment by willothy** [comment, marked as answer]
I just tried this in my config and it does work.

Just replace
```lua
vim.schedule_wrap(function()
  vim.cmd.startinsert()
end)
```
with
```lua
vim.schedule(vim.cmd.startinsert)
```

in your `post_open` callback.

**Reply by przepompownia** [reply]
Indeed, `vim.schedule_wrap` returns a function only and does not execute it so using it there does not make sense. Thanks!

**Reply by przepompownia** [reply]
Btw. in the meantime I found some (ugly) way to avoid using nesting nvim instance for making a commit: https://github.com/przepompownia/git-utils.nvim

**Reply by willothy** [reply]
True! You could do that. I actually use flatten to open git commits in different wezterm tabs and auto-switch back, so doesn't quite fit my use case, but definitely good to know :)
